meta-iotqa/lib/oeqa/selftest/computervision-no-gplv3.py

```diff
@@ -0,0 +1,423 @@
+    rRundeps = [getPackage(package) for package in rundeps]
+
+    rdepsCache[package] = rRundeps
+    return rRundeps
+
+def getLicenses(package):
+    if package in licenseCache:
+        return licenseCache[package]
+
+    licenseProp = subprocess.check_output(["oe-pkgdata-util", "read-value", "LICENSE", package]).decode("utf-8")
+
+    licenseCache[package] = licenseProp
+    return licenseProp
+
+def findChildren(name, whiteList, chain=[]):
+    results = []
+
+    rundeps = getRdeps(name)
+
+    for d in rundeps:
+        recipe = getRecipe(d)
+        if recipe in whiteList:
+            # Do not process whitelisted dependencies.
+            continue
+        if d in chain:
+            # Take away possible loops.
+            continue
+
+        # print(str(chain) + " -> " + d + ": " + str(rundeps))
+
+        licenses = parseLicenseString(getLicenses(d))
+
+        children = findChildren(d, whiteList, chain + [d])
+        childNode = LicenseNode(licenses, children, d)
+
+        results.append(childNode)
+
+    return results
+
+def createTree(package, whiteList=[]):
+
+    rundeps = getRdeps(package)
+
+    licenses = parseLicenseString(getLicenses(package))
+
+    children = findChildren(package, whiteList, [package])
+    root = LicenseNode(licenses, children, package)
+
+    return root
+
+class GplV3Test(oeSelfTest):
+    """GPLv3 licensing test class."""
+
+    test_image = 'refkit-image-computervision-nogplv3'
+
+    @classmethod
+    def setUpClass(cls):
+        # Create the test image. We don't know which packages end up to
+        # the image without building it, because manifes isn't available
+        # with just "bitbake -c rootfs".
+        print('Building test image (%s)...' % cls.test_image)
+        bitbake('%s' % cls.test_image)
+
+    @classmethod
+    def tearDownClass(cls):
+        print('Cleaning...')
+        bitbake('-c clean %s' % cls.test_image)
+
+    def test_check_licensing(self):
+        """ Check that licenses are possible without using GPLv3 family licenses in any component """
+        print("test_check_licensing")
+
+        whiteList = []
+        packageNames = []
+        lines = None
+
+        bb_vars = get_bb_vars(["DEPLOY_DIR_IMAGE", "IMAGE_LINK_NAME", "META_REFKIT_BASE"], self.test_image)
+        deploydir = bb_vars["DEPLOY_DIR_IMAGE"]
+        imagename = bb_vars["IMAGE_LINK_NAME"]
+        basedir = bb_vars["META_REFKIT_BASE"]
+        manifest = os.path.join(deploydir, imagename + ".manifest")
+        whitelist = os.path.join(basedir, "../meta-iotqa/scripts/contrib/licensetree-whitelist.txt")
+
+        self.assertTrue(os.path.islink(manifest), msg="No manifest file created for image. It should have been created in %s" % manifest)
+
+        # Whitelist contains the runtime dependencies which do not
+        # propagate licenses. For example when a component uses a D-Bus
+        # API or execs another process.
```

Comment:
The whitelist might have some components which are not present (I tested this on Poky too). Another way to approach this would be to have a "blacklist" of library components which can cause license propagation. I checked the `SECTION` variable but it appears to be unrealiable: few recipes do package-level assignment to sections, some libraries are assigned to `devel` or `utils` instead of `libs`, and so on.

Since adding more licenses to the runtime dependency tree should only add constraints, so having a component forgotten from the whitelist would only result in possible false positives, meaning that the test will fail when it should have succeeded.